presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDictionarySelectiveReader.java

```diff
@@ -384,32 +388,41 @@ public Block getBlock(int[] positions, int positionCount)
             return new RunLengthEncodedBlock(outputType.createBlockBuilder(null, 1).appendNull().build(), positionCount);
         }
 
-        wrapDictionaryIfNecessary();
+        // compact values(ids) array, and calculate 1) the slice sizeInBytes if materialized, and 2) number of nulls
```

Comment:
Compaction logic is somewhat difficult to get right, hence, it would be better to re-use it by calling compactValues. Also, blockSizeInBytes and nullCount can be computed in separate loops for readability.

```
        compactValues(positions, positionCount);

        long blockSizeInBytes = 0;
        for (int i = 0; i < positionCount; i++) {
            int id = values[i];
            blockSizeInBytes += dictionaryOffsetVector[id + 1] - dictionaryOffsetVector[id];
        }

        int nullCount = 0;  // the nulls count for selected positions
        for (int i = 0; i < positionCount; i++) {
            int id = values[i];
            nullCount += (id == currentDictionarySize - 1 ? 1 : 0);
        }
```

Reply:
@mbasmanova Thanks for the rewrite! How about calculating the in the same loop? This will use less number of lines and still readable:
```
        compactValues(positions, positionCount);

        long blockSizeInBytes = 0;
        int nullsCount = 0;
        for (int i = 0; i < positionCount; i++) {
            int id = values[i];
            blockSizeInBytes += dictionaryOffsetVector[id + 1] - dictionaryOffsetVector[id];
            nullsCount += (id == currentDictionarySize - 1 ? 1 : 0);
        }
```

Reply:
That works as long as we get to reuse compactValues. Also, null count calculation can be mode a bit more readable:

```
if (id == currentDictionarySize - 1) {
   nullCount++;
}
```